### DISCUSSION TITLE
HoverTool model callback property

### DISCUSSION THREAD

**Discussion by gmerritt123** [bokeh/bokeh#12739] [Internals and design (Q&A)] [answered]
Hi, trying to wrap my head around the HoverTool model's callback property, specifically the reasoning behind limiting it to one callback:

Like, why is it:

```
hvr = HoverTool()
hvr.callback = CustomJS().
```

and not something like:

```
hvr.callbacks = [CustomJS(), ...]
```

or even why does it exist at all, and not be something event based, like

```
hvr = HoverTool()
hvr.js_on_event('hover',CustomJS()) #no such thing as hover event
#... then i can add more callbacks to it

# or even
hvr.js_on_hover(CustomJS()) #kinda like js_on_click for Buttons for example
```

The intricacies/possibilities are pretty overwhelming and I don't know exactly what the best way forward might be, or if there's already a workaround i'm missing. 

Having access to cb_data and cb_obj in each callback really shouldn't matter --> that would just be the same thing in each CustomJS that's attached.

Then I thought of workarounds. If there's no hover event", maybe I can use js_on_change? Then that reminded me of the classic "selection" callback question that gets asked all the time, where the nuance is that you don't trigger CustomJS on selection event, but instead trigger it when the selected indices of the underlying datasource changes--> e.g.  `src.selected.js_on_change('indices',....)`.

But there is no `src.inspected` exposed for me to do that! (saw this for example https://github.com/bokeh/bokeh/issues/8792)

So I guess I'm kinda paralyzed with the options. Is there a workaround/other option I'm missing? Is a feature request in order for this, and if so, what should that look like?

As always, thanks so much for all your work.

**Comment by bryevdv** [comment]
> why is it

History, no other reason. Properties for `.callback` were the first stab at adding callback capabilities, about 10 years ago. Almost al those early `.callback` APIs are gone now, with 1 or 2 exceptions, including the one above in particular. It should really go away, just no one has gotten around to it yet. 

IIRC the thing that `Hovertool.callback` does that has not been replicated in a new/different way yet is that it fires on every mouse move, whether a "hit" happens or not. `js_on_event('hover', ...)` might be a plausible way to replace that in a more modern way. A `source.inspected`  would be a way to get events, but only on a hit. But someone will need to go do a little but of analysis and planning ahead. 

Think the question about workarounds is specifically for multiple callbacks on one hover? I don't have any especially good suggestions at present. I suppose you could stuff two (or however many) distinct `CustomJS` instances the `args` dict of a "top level" `CustomJS` and then manually invoke the `execute` of all the other CustomJS there. But that seems fragile and hacky, and there may even be general problems preventing it being successful, that I'm not aware of off the top of my head (would just have to try)

cc @bokeh/core

**Comment by gmerritt123** [comment, marked as answer]
Thanks for the response. Yeah, chaining callbacks gets messy in a hurry. 

FYI my workaround was simply to add js_on_event('mousemove',cb) to the figure holding the hovertool, and then cb was something like:

```
cb=CustomJS(args= dict( src=src,...), code=```
if (src.inspected.indices.length>0){
    //do stuff
    }
```

I figured since the Hovertool.callback hits on every mousemove anyway, I might as well just mimic it on the figure with an if statement checking for inspection. This works fine, I can add n callbacks this way without needing them executing inside each other.

I do agree it'd be good to modernize the HoverTool callback, one way or another.
